near-sdk/src/store/tree_map/iter.rs

```diff
@@ -507,6 +638,98 @@ where
+                } else {
+                    seen = Some(&node.key);
+                    self.stack_asc.push(*curr_idx);
+                    curr = node.lft.as_ref();
+                }
+            } else {
+                curr = None
+            }
+        }
+        seen
+    }
+
+    //The last element in the stack is the last returned key.
+    //Find the next key to the last item in the stack
+    fn find_next_asc(&mut self) -> Option<&'a K> {
+        let last_key_idx = self.stack_asc.pop();
+        let mut seen: Option<&K> = None;
+        if let Some(last_idx) = last_key_idx {
+            if let Some(node) = self.tree.node(last_idx) {
+                //If the last returned key has right node then return minimum key from the
+                //tree where the right node is the root.
+                seen = match node.rgt {
+                    Some(rgt) => self.find_min(Some(&rgt), Bound::Unbounded),
+                    None => None,
+                }
+            }
+        }
+        //If the last returned key does not have right node then return the
+        //last value in the stack.
+        if seen.is_none() && !self.stack_asc.is_empty() {
+            if let Some(result_idx) = self.stack_asc.last() {
+                seen = self.tree.node(*result_idx).map(|f| &f.key);
+            }
+        }
+        seen
+    }
+    fn find_max(&mut self, root: Option<&FreeListIndex>, max: Bound<&'a K>) -> Option<&'a K> {
+        let mut curr = root;
+        let mut seen: Option<&K> = None;
+
+        while let Some(curr_idx) = curr {
+            if let Some(node) = self.tree.node(*curr_idx) {
+                if key_gt_bound(&node.key, max) {
+                    curr = node.lft.as_ref();
+                } else {
+                    seen = Some(&node.key);
+                    self.stack_desc.push(*curr_idx);
+                    curr = node.rgt.as_ref();
+                }
+            } else {
+                curr = None
+            }
+        }
+        seen
+    }
+
+    fn find_next_desc(&mut self) -> Option<&'a K> {
+        let last_key_idx = self.stack_desc.pop();
+        let mut seen: Option<&K> = None;
+        if let Some(last_idx) = last_key_idx {
+            if let Some(node) = self.tree.node(last_idx) {
+                //If the last returned key has left node then return maximum key from the
+                //tree where the left node is the root.
+                seen = match node.lft {
+                    Some(lft) => self.find_max(Some(&lft), Bound::Unbounded),
+                    None => None,
+                }
+            }
+        }
+        //If the last returned key does not have left node then return the
+        //last value in the stack.
+        if seen.is_none() && !self.stack_desc.is_empty() {
+            if let Some(result_idx) = self.stack_desc.last() {
+                seen = self.tree.node(*result_idx).map(|f| &f.key);
+            }
+        }
+        seen
+    }
+}
```

Comment:
At a glance, this logic seems identical to that for `Keys`. What is preventing it from being abstracted to be shared between the two?

Reply:
`Keys` seems to be special case of `KeysRange` and can be easily replaced. May be a followup Issue to remove `Keys` and replace with `KeysRange` might be a better solution. What do you think?

Reply:
yeah, that's fine, can do in a follow up!

@itegulov feel free to merge this if you're happy with it, unsure if you had any other thoughts since you seem to be the stakeholder for that issue

Reply:
Created #987 to track the issue.